### PR TITLE
Create a component guide based on components in an application

### PR DESCRIPTION
**Pull request by fofr**
Part of:
https://trello.com/c/SD6M3IUF/12-3-create-a-component-guide-that-can-document-components-in-apps

Not all components should be in static. If they are used only in one application it is preferable to keep them close to the point of use.

Initial release:
* Allow applications to create components
* Components follow the same file structure as static components
* Component guide is generated from docs and partials living in `app/views/components/docs` and `app/views/components`
* An example of how to use this gem: https://github.com/alphagov/government-frontend/compare/components-gem
* Component guide logic, templates and styles ported from https://github.com/alphagov/govuk-component-guide and http://govuk-component-guide.herokuapp.com/
* Includes basic integration tests

See README.md for further details.

Still to do but can be separate PRs:
* Syntax highlighting
* Component guide styles should be components themselves
* Further documentation on components and principles
* Component previews that render just the component (ie no GOV.UK frame around it)

screen shot 2017-07-19 at 17 18 31
screen shot 2017-07-19 at 17 18 49

### CODE REVIEW

File: Gemfile.lock [hunk 1]
```diff
@@ -2,7 +2,10 @@ PATH
   remote: .
```
Comment (tijmenb) [inline Gemfile.lock]: The `Gemfile.lock` shouldn't be committed for gems.

Comment (fofr) [inline Gemfile.lock]: Thanks, removed in 32895dd

File: README.md [hunk 2]
```diff
@@ -16,13 +15,63 @@ And then execute:
...
+`govuk_publishing_components` expects the following components structure in your application:
+
+* a Partial View in `app/views/components` eg `_component-name.html.erb` - The template logic and markup, also defines the arguments expected
+* Documentation in `/app/views/components/docs` - a `.yml` per component, describing the component and containing fixture data.
+* a SCSS module in `app/assets/stylesheets/components` - The styling of the component
+* a Javascript module in `app/assets/javascripts/components` - JS behaviour of the component
+* a unit test in `/test/components` - testing the component renders correctly based on parameters passed in
+
+The component guide will include your application’s styles and scripts. It will include `application.scss` and `application.js` by default but this is [configurable](#configuration).
+
```
Comment (andysellick) [inline README.md]: Should there be a line or two in here about how to call/use a component? I guess it's already going to be in the component guide though.

File: README.md [hunk 3]
```diff
@@ -1,11 +1,10 @@
...
+# GOV.UK Publishing Components
+
+A gem to document components in GOV.UK frontend applications.
 
 ## Usage
-How to use my plugin.
 
-## Installation
-Add this line to your application's Gemfile:
+Add this line to your application's Gemfile in the development and test groups:
```
Comment (andysellick) [inline README.md]: As someone new to this, I don't know what the "development and test groups" are, so I'd be stuck on this step.

File: README.md [hunk 4]
```diff
@@ -16,13 +15,63 @@ And then execute:
...
+### Mount the component guide
+
+Mount the component guide in your application:
+
+```ruby
+# config/routes.rb
+unless Rails.env.production?
+  mount GovukPublishingComponents::Engine, at: "/component-guide"
+end
+```
```
Comment (andysellick) [inline README.md]: Is this a change that once done is okay to commit to that application? If so, presumably over time eventually all apps will have this already. Some clarification/expansion needed here?

File: README.md [hunk 5]
```diff
@@ -16,13 +15,63 @@ And then execute:
...
+
+For example, if your application was government-frontend, the component guide would be found at:
+`government-frontend.dev.gov.uk/component-guide`
+
+### Write components
+
+`govuk_publishing_components` expects the following components structure in your application:
+
+* a Partial View in `app/views/components` eg `_component-name.html.erb` - The template logic and markup, also defines the arguments expected
```
Comment (andysellick) [inline README.md]: Having an example file name here is helpful - can example file names also be added for the following items as well?

File: README.md [hunk 6]
```diff
@@ -16,13 +15,63 @@ And then execute:
...
+* Documentation in `/app/views/components/docs` - a `.yml` per component, describing the component and containing fixture data.
+* a SCSS module in `app/assets/stylesheets/components` - The styling of the component
+* a Javascript module in `app/assets/javascripts/components` - JS behaviour of the component
+* a unit test in `/test/components` - testing the component renders correctly based on parameters passed in
```
Comment (vanitabarrett) [inline README.md]: Is it worth pointing people towards an example component somewhere, so they can see exactly what you mean by the above points? Things like 'defines the arguments expected' - it may not be immediately obvious to everyone exactly how this is done?

Comment (vanitabarrett) [inline README.md]: Also, even though we're not giving people a comprehensive guide on how to write a component here, it might still be worth linking to our component principles?

File: README.md [hunk 7]
```diff
@@ -16,13 +15,63 @@ And then execute:
 $ bundle
 ```
```
Comment (andysellick) [inline README.md]: Where should these commands be executed?

File: README.md (hunk 2 above)
Comment (fofr) [inline README.md]: I think it's worth adding one or two lines.

File: README.md [hunk 8]
```diff
@@ -16,13 +15,63 @@ And then execute:
...
+description: A banner that displays when content is withdrawn
+fixtures:
+  default:
+    title: 'This consultation was withdrawn on 20 April 2017'
+    explanation_html: '<p>A supporting paragraph</p>'
+```
+
+### Configuration
+
+You can configure the gem with a config block in an initializer:
```
Comment (andysellick) [inline README.md]: Remove 'you can'.

File: README.md (hunk 6 above)
Comment (fofr) [inline README.md]: Yes, definitely. Holding off doing this until we have one to link to.

File: README.md (hunk 7 above)
Comment (fofr) [inline README.md]: These are the standard Rails "How to install gem" instructions.

**Comment by fofr** [comment]
README updated in https://github.com/alphagov/govuk_publishing_components/pull/1/commits/b929dd2aced28550096f60eebdbd6b7fe6bdf17a

File: README.md [hunk 9]
```diff
@@ -16,13 +15,75 @@ And then execute:
...
+
+The `.yml` file must have:
+* a name
+* a description
+* a default fixture (this can be `{}` if the component takes no parameters)
+
+#### Example documentation
+
+```yaml
+name: Withdrawal notice
```
Comment (alecgibson) [inline README.md]: (Just a thought, but it might be nice with all of these examples to demonstrate a single component? Then you can see the process all the way through for one thing. Above for your `<%= render` example, you use descriptions; for the path examples you use a print link; here you're using a withdrawal notice)

File: README.md [hunk 10]
```diff
@@ -16,13 +15,75 @@ And then execute:
...
+* test/components/print_link_test.rb
+
+The component guide will include your application’s styles and scripts. It will include `application.scss` and `application.js` by default but this is [configurable](#configuration).
+
+### Write documentation
+
+The `.yml` file must have:
+* a name
+* a description
+* a default fixture (this can be `{}` if the component takes no parameters)
```
Comment (alecgibson) [inline README.md]: It might be nice to expand on what a "default" fixture is? And what a fixture that isn't default is?

File: README.md [hunk 11]
```diff
@@ -16,13 +15,75 @@ And then execute:
...
+```
+
+### Configuration
+
+Configure the gem with a config block in an initializer:
+
+```ruby
+# config/initializers/govuk_publishing_components.rb
+GovukPublishingComponents.configure do |c|
+  c.component_guide_title = "My component guide"
```
Comment (alecgibson) [inline README.md]: What's the use case for being able to configure this?

File: app/controllers/govuk_publishing_components/component_guide_controller.rb [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+                             {
+                               title: @component_doc.name
+                             }
+                           ]
+    end
+
+    def fixture
+      @component_doc = ComponentDoc.get(params[:component])
+      @component_fixture = @component_doc.fixtures.find { |f| f.id == params[:fixture] }
+      @guide_breadcrumbs = [
```
Comment (alecgibson) [inline app/controllers/govuk_publishing_components/component_guide_controller.rb]: It might be nice to break these breadcrumbs out into their own little method.

File: app/models/govuk_publishing_components/component_doc.rb [hunk 13]
```diff
@@ -0,0 +1,34 @@
+module GovukPublishingComponents
+  ComponentDoc = Struct.new(:id, :name, :description, :body, :fixtures) do
```
Comment (alecgibson) [inline app/models/govuk_publishing_components/component_doc.rb]: This class has a lot of collection-level functionality, which feels like it should belong in another separate singleton class that has the ability to memoize things properly, and deliver things from hashes rather than performing `find` operations.

File: app/models/govuk_publishing_components/component_doc.rb [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+    def fixture
+      fixtures.first
+    end
+
+    def other_fixtures
+      fixtures.slice(1..-1)
+    end
+
+    def self.fetch_component_doc
```
Comment (alecgibson) [inline app/models/govuk_publishing_components/component_doc.rb]: fetch_component_doc**s** (plural)?

File: app/models/govuk_publishing_components/component_fixture.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+module GovukPublishingComponents
+  ComponentFixture = Struct.new(:id, :data) do
+    def name
+      id.humanize
+    end
+
+    def pretty_data
+      JSON.pretty_generate(data).gsub('\\n', "\n    ").gsub(/"(\w*)":/, '\1:').strip
```
Comment (alecgibson) [inline app/models/govuk_publishing_components/component_fixture.rb]: When performing this sort of `gsub` wizardry, it's often nice to break it out into a one-line mutating method with a nice name so that people can understand what's going on, eg:

```ruby
def pretty_data
  prettified = JSON.pretty_generate(data)
  prepend_new_lines_with_whitespace!(prettified)
  unquote_properties!(prettified)
  prettified
end
```

On that note, just why _are_ we doing these things? To make a Ruby-looking hash? In that case maybe we should be using a gem? Or at least rename this method to something like `data_as_pretty_printed_ruby_hash`?

File: app/models/govuk_publishing_components/component_doc.rb [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    def self.build(component)
+      fixtures = component[:fixtures].map { |id, data|
+        GovukPublishingComponents::ComponentFixture.new(id.to_s, data)
+      }
+
+      self.new(component[:id], component[:name], component[:description], component[:body], fixtures)
+    end
+
+    def fixture
+      fixtures.first
```
Comment (alecgibson) [inline app/models/govuk_publishing_components/component_doc.rb]: This doesn't necessarily grab the "default" fixture - is that what we want?

File: app/views/govuk_publishing_components/component_guide/show.html.erb [hunk 17]
```diff
@@ -0,0 +1,32 @@
+<% content_for :title, "#{@component_doc.name} component" %>
+<%= render 'govuk_component/title', title: @component_doc.name, context: 'Component' %>
+
+<p class="component-description">
+  <%= @component_doc.description %>
+</p>
+
+<div class="component-show">
+  <div class="component-doc">
+    <h3>How to call this component</h3>
```
Comment (alecgibson) [inline app/views/govuk_publishing_components/component_guide/show.html.erb]: (This view and the `fixture` view share a bit of code in common. I don't know if it's possible to commonise through a partial?)

File: README.md (hunk 10 above)
Comment (fofr) [inline README.md]: I agree. My intention is to fix this, so a component without any fixtures does not take parameters. And an "empty parameter" option can be easily illustrated. Was going to do this as a separate PR.

File: README.md (hunk 9 above)
Comment (fofr) [inline README.md]: Good point. When we're written some I'll link directly to the files in other apps.

File: README.md (hunk 11 above)
Comment (fofr) [inline README.md]: Unique titles for component guides in different apps.

File: app/models/govuk_publishing_components/component_doc.rb (hunk 13 above)
Comment (fofr) [inline app/models/govuk_publishing_components/component_doc.rb]: This logic has largely been ported from https://github.com/alphagov/govuk-component-guide/

https://github.com/alphagov/govuk-component-guide/blob/master/app/models/component.rb

File: app/models/govuk_publishing_components/component_fixture.rb (hunk 15 above)
Comment (fofr) [inline app/models/govuk_publishing_components/component_fixture.rb]: Another ported model from here:
https://github.com/alphagov/govuk-component-guide/blob/master/app/models/fixture.rb#L7

It creates a hash that can be put into a `<pre>` block with correct formatting. I agree the naming should be better.

File: README.md (hunk 11 above)
Comment (fofr) [inline README.md]: See: https://github.com/alphagov/government-frontend/pull/403/files#diff-bb6adab3eaf25dfc303b4f34455f5be9R2

File: app/controllers/govuk_publishing_components/component_guide_controller.rb (hunk 12 above)
Comment (fofr) [inline app/controllers/govuk_publishing_components/component_guide_controller.rb]: Added in 84537ad

File: app/models/govuk_publishing_components/component_doc.rb (hunk 14 above)
Comment (fofr) [inline app/models/govuk_publishing_components/component_doc.rb]: Fixed in d723a47

File: README.md (hunk 11 above)
Comment (alecgibson) [inline README.md]: I understand what it does, but why would we want to do this?

Comment (fofr) [inline README.md]: Because each guide will be different. Each guide will have different components.

File: app/models/govuk_publishing_components/component_doc.rb (hunk 16 above)
Comment (fofr) [inline app/models/govuk_publishing_components/component_doc.rb]: This relies on convention, you’re right. I think this is ok for now. The convention of putting default first in the fixtures hasn't been an issue with static components.

The alternative would be to lookup via the `default` ID – if this was left out what should the guide page do? At the moment it'd fallback to the next fixture.

I think when there's a component generator this convention will be natural too.

**Comment by fofr** [comment]
Plan regarding logic and review comments is to:
* Integrate this gem into the govuk-component-guide and remove duplicate logic: https://trello.com/c/GtIEFBfC/
* Then refactor
